### PR TITLE
Haxe 4 changes, part 2

### PR DESCRIPTION
**Pull request by Aurel300**
Continuing #383 for an easier review process.

## Syntax

 - [x] ~~XML~~ Markup literals

## Stdlib

 - [x] `implements Dynamic` on `extern`s only
 - [x] `sys.thread`
 - ~~`haxe.iterators`~~ (I mentioned the Unicode iterator, the others are more or less part of other APIs anyway)

## Compiler defines

 - [x] `warn_var_shadowing`
 - ~~update the list of defines~~ (#389)

## Compiler features

 - [x] null safety

## Target changes

 - [x] `*.Syntax` classes to replace `untyped __{js,php,...}__`
 - JS
   - [x] ~~ES6 generation via `-D js-es=6`~~ (already documented)
   - [x] externs for `Map`, `Set`, `Date` ...
   - [x] some externs moved to `js.lib`
 - PHP
   - [x] minimum version supported is now PHP 7.0
   - [x] `--php-{prefix,front,lib}` changed to `-D php-*=`
 - Lua
   - [x] `-D lua-vanilla`
 - macro
   - [x] `eval` target ?
   - [x] debugger
   - [x] `@:persistent` (replacing `registerModuleReuseCall` and `onMacroContextReused`)

## External libraries

 - [x] jQuery and swfobject
 - [x] `hx3compat`
   - `haxe.remoting`
   - `haxe.web.Request`
   - `haxe.unit`
 - ~~`record-macros`~~ (not documented in the first place)

## Other changes

 - [x] disallow macro-in-macro
 - [x] Unicode support
 - [x] arguments consistently use `--long-argument` now
 - [x] `.<platform>.hx` files
 - [x] `@:using`
 - [x] mention `display` + IDE integration, `vshaxe` ?
 - ~~enum values without arguments as default function argument values~~ (#386 + https://github.com/HaxeFoundation/haxe/issues/8187)

### CODE REVIEW

File: HaxeManual/12-target-details.tex [hunk 1]
```diff
@@ -253,6 +253,10 @@ \subsection{ECMAScript 6 class generation}
 \subsection{Using external JavaScript libraries}
 \label{target-javascript-external-libraries}
 
+\since{4.0.0}
+
+\emph{Deprecated:} The jQuery and SWFObject externs have been removed from the standard library in Haxe 4. For compatibility purposes they are available in the \href{https://github.com/HaxeFoundation/hx3compat/}{hx3compat} library.
```
Comment (Gama11) [inline HaxeManual/12-target-details.tex]: Missing `\type{}`?

Comment (Aurel300) [reply HaxeManual/12-target-details.tex]: (per Slack chat, these refer to the library names, no reason to `\type`)

File: HaxeManual/10-std.tex [hunk 2]
```diff
@@ -1055,6 +1059,10 @@ \subsection{Implementation details}
 \section{Unit Testing}
 \label{std-unit-testing}
 
+\since{4.0.0}
+
+\emph{Deprecated:} The unit testing API has been removed from the standard library in Haxe 4. For compatibility purposes it is still available in the \href{https://github.com/HaxeFoundation/hx3compat/}{hx3compat} library. See \Fullref{std-hx3compat}.
```
Comment (Gama11) [inline HaxeManual/10-std.tex]: I wonder if alternatives like utest should be recommended here?

Comment (Gama11) [reply HaxeManual/10-std.tex]: Oh, nvm, I just read further... :)

File: HaxeManual/10-std.tex [hunk 3]
```diff
@@ -1128,3 +1136,28 @@ \section{Unit Testing}
...
+	\item \type{haxe.unit.*} Unit testing
+	\item \type{haxe.web.*} Web request classes for Haxe web servers
+	\item \type{neko.net.*} Neko web server
+	\item (JavaScript target only) jQuery and SWFObject externs
+\end{itemize}
+
+Note that some of these APIs were removed form the standard library in favour of solutions in external libraries:
+
+\begin{itemize}
+	\item \href{https://github.com/haxe-utest/utest}{haxe-utest} is a Haxe unit testing library.
```
Comment (Gama11) [inline HaxeManual/10-std.tex]: The haxe prefix seems slightly redundant? The lib name is just `utest`, not `haxe-utest`.

Comment (Aurel300) [reply HaxeManual/10-std.tex]: Oh yeah, it's the GitHub org name that is `haxe-utest`.

File: HaxeManual/06-language-features.tex [hunk 4]
```diff
@@ -455,6 +455,22 @@ \subsection{In the Haxe Standard Library}
...
+
+\paragraph{Exclusion from Static Extension}
+
+By default, using a class as a static extension brings all of its static methods into the context, allowing them to be used as extensions of the appropriate types. In certain situations, the class can provide other static methods which are not intended for static extension. To make sure they do not interfere with the proper methods of the type, these methods can be marked with \expr{@:noUsing}:
+
+\haxe{assets/StaticExtension3.hx}
+
+\paragraph{Default Static Extension}
+
+It is also possible to always enable particular static extensions for a given type, by annotating the type with the \expr{@:using(args...)} metadata. The arguments are the full dot paths of static extension classes that will be applied on the type:
```
Comment (Gama11) [inline HaxeManual/06-language-features.tex]: This implies that `@:using` works for every type, while it seems to basically only work for enums and classes at the moment...

https://github.com/HaxeFoundation/haxe/issues/8022
https://github.com/HaxeFoundation/haxe/issues/7930
https://github.com/HaxeFoundation/haxe/issues/7859

Comment (Aurel300) [reply HaxeManual/06-language-features.tex]: I guess we'll leave this as-is for now :)

File: HaxeManual/13-debugging.tex [hunk 5]
```diff
@@ -68,6 +68,12 @@ \section{Tracing Types}
 \end{lstlisting}
 
 
+\section{IDE Integration}
+\label{debugging-ide-integration}
+
+Haxe is supported in a \href{https://haxe.org/documentation/introduction/editors-and-ides.html}{number of editors and IDEs}. Some IDEs, such as Visual Studio Code (with the \href{https://marketplace.visualstudio.com/items?itemName=nadako.vshaxe}{vshaxe} extension), support debugging of Haxe code with breakpoints, stack tracing, etc. Simply run the program in debug mode.
```
Comment (Gama11) [inline HaxeManual/13-debugging.tex]: vshaxe itself only has debugging for eval built-in, so it may be misleading to link to the marketplace page for that. Perhaps https://github.com/vshaxe/vshaxe/wiki/Debugging is a better thing to link to.

File: HaxeManual/03-type-system.tex [hunk 6]
```diff
@@ -576,3 +576,11 @@ \subsection{Resolution Order}
...
 	\item For \tref{static extensions}{lf-static-extension}, a match is made if the name equals and the first argument \tref{unifies}{type-system-unification}. Within a given type being used as static extension, the fields are checked from top to bottom.
 \end{itemize}
+
+
+\section{untyped}
+\label{type-system-untyped}
+
+\emph{Important note:} This syntax should be avoided whenever possible. The produced code cannot be properly checked by the Haxe compiler and so it may have type errors or other bugs that would be caught at compile time in regular code. Use only when absolutely necessary and when you know what you are doing.
+
+It is possible to completely circumvent the type checker by prefixing an expression with the keyword \expr{untyped}. Any expression that can be parsed is then accepted and emitted into the generated output. This is primarily used with the target-specific \tref{code injection expressions}{target-syntax}.
```
Comment (Simn) [inline HaxeManual/03-type-system.tex]: That's not quite true, there are still some typing errors from which the compiler doesn't recover.

File: HaxeManual/06-language-features.tex [hunk 7]
```diff
@@ -165,12 +165,25 @@ \chapter{Language Features}
...
+    new Main<String>("foo");
+    new Main(12); // use type inference
+  }
 
-    function new(a:A) { }
+  function new(a:A) { }
+}
+\end{lstlisting}
+
+\emph{\tref{Markup Literals}{lf-markup}:}
```
Comment (Simn) [inline HaxeManual/06-language-features.tex]: Let's not add this for now.

File: HaxeManual/06-language-features.tex [hunk 8]
```diff
@@ -218,6 +231,74 @@ \section{Conditional Compilation}
...
 \end{itemize}
 
+\subsection{Target defines}
+\label{lf-target-defines}
+
+Depending on the current target, one of the following flags will be defined. Note that they are the same as the \tref{argument provided to the compiler to specify the output}{compiler-usage}.
+
+\begin{itemize}
+	\item[\ic{as3}] Defined when generating ActionScript 3 code.
+	\item[\ic{cpp}] Defined when generating C++ code.
```
Comment (Simn) [inline HaxeManual/06-language-features.tex]: It is also defined for `cppia` output.

File: HaxeManual/06-language-features.tex [hunk 9]
```diff
@@ -218,6 +231,74 @@ \section{Conditional Compilation}
...
+	\item[\ic{cppia}] Defined when generating a cppia script.
+	\item[\ic{cs}] Defined when generating C\# code.
+	\item[\ic{eval}] Defined when running the code with \ic{--interp}.
```
Comment (Simn) [inline HaxeManual/06-language-features.tex]: It is also defined in macros.

File: HaxeManual/09-macros.tex [hunk 10]
```diff
@@ -298,6 +298,15 @@ \section{Limitations}
 \subsection{Macro-in-Macro}
 \label{macro-limitations-macro-in-macro}
 
+Build macros cannot be invoked from a macro context. This means it is impossible to use a macro to create a class which will provide build macros for other classes.
+
+\since{4.0.0}
+
+Expression macros are also disallowed in a macro context.
```
Comment (Simn) [inline HaxeManual/09-macros.tex]: I think we should mention that they only fail (i.e. throw an exception) if they are actually invoked. This is important because it avoids a lot of fencing.

File: HaxeManual/09-macros.tex [hunk 11]
```diff
@@ -325,10 +334,45 @@ \subsection{Type Parameters}
...
+\end{lstlisting}
+
+After starting the completion server with \ic{haxe --wait 6000}, we perform a build with \ic{haxe --connect 6000 --main Main --no-output}. The first time the congratulatory message is printed during compilation. If we execute the same command again, however, it is not - the \expr{firstBuild} variable retains the value \expr{false} from the previous build.
+
+
 \section{Initialization Macros}
 \label{macro-initialization}
 
-Initialization macros are invoked from command line by using the \expr{--macro callExpr(args)} command. This registers a callback which the compiler invokes after creating its context, but before typing what was argument to \expr{-main}. This then allows configuring the compiler in some ways.
+Initialization macros are invoked from command line by using the \expr{--macro callExpr(args)} command. This registers a callback which the compiler invokes after creating its context, but before typing what was argument to \expr{--main}. This then allows configuring the compiler in some ways.
```
Comment (Simn) [inline HaxeManual/09-macros.tex]: You use `--main` here but `-m` somewhere above. Personally, I still prefer `-main`, but I guess we should at least be consistent.

File: HaxeManual/10-std.tex [hunk 12]
```diff
@@ -7,15 +7,22 @@ \section{String}
...
+Use the \ic{.code} property on a constant single-char string in order to compile its Unicode character point:
 
 \begin{lstlisting}
 "#".code // will compile as 35
 \end{lstlisting}
 
+\since{4.0.0}
+
+\paragraph{Unicode support}
+All Haxe targets except Neko support Unicode in strings by default. The compiler define \expr{target.unicode} is defined when the current target supports Unicode.
```
Comment (Simn) [inline HaxeManual/10-std.tex]: That define doesn't exist because Nicolas didn't want it.

File: HaxeManual/10-std.tex [hunk 13]
```diff
@@ -7,15 +7,22 @@ \section{String}
...
+
+A string in Haxe code represents a valid sequence of Unicode codepoints. Due to differing internal representations of strings across targets, only the basic multilingual plane (BMP) is supported consistently: every BMP Unicode codepoint corresponds to exactly one string character.
+
+On some targets, the internal representation is UTF-16, which means that non-BMP Unicode codepoints are represented using surrogate pairs. It is still possible to work with strings on these targets without having to manually decode surrogate pairs by using the \href{https://api.haxe.org/v/development/haxe/iterators/StringIteratorUnicode.html}{Unicode iterators API} provided in the standard library.
```
Comment (Simn) [inline HaxeManual/10-std.tex]: I thought that we wanted to remove StringIteratorUnicode in favor of Utf8.iterate, but I'm not sure. Have to keep that in mind though.

File: HaxeManual/10-std.tex [hunk 14]
```diff
@@ -868,12 +875,37 @@ \subsection{Implementation details}
...
+\subsection{Threading}
+\label{std-threading}
+
+\since{4.0.0}
+
+A \href{https://api.haxe.org/sys/thread/Thread.html}{unified threading API} is available on some sys targets. The compile-time define \expr{target.threaded} is set when the API is available. The API allows very simple creation of threads from functions:
+
+\haxe{assets/Threads.hx}
+
+All spawned threads are currently treated as daemon threads, meaning that the main thread will not wait for their completion.
```
Comment (Simn) [inline HaxeManual/10-std.tex]: I'd lose the "currently". I don't see us changing this.

File: HaxeManual/10-std.tex [hunk 15]
```diff
@@ -868,12 +875,37 @@ \subsection{Implementation details}
...
+
+Due to the nature of threads accessing a shared memory space with all the Haxe variables and objects, it is possible to run into issues due to deadlocks and race conditions. The standard library provides some core synchronization constructs in the \href{https://api.haxe.org/sys/thread/}{\expr{sys.thread}} package.
```
Comment (Simn) [inline HaxeManual/10-std.tex]: Phrasing is a bit weird, sounds like the nature of threads is doing something.

File: HaxeManual/12-target-details.tex [hunk 16]
```diff
@@ -474,13 +429,13 @@ \subsection{Exposing Haxe classes for JavaScript}
 
 In this pattern, a var statement is used to expose the module; it doesn't write to the \ic{window} or \ic{exports} object. 
 
-\subsection{Loading extern classes using "require" function}
+\subsection{Loading extern classes using ``require'' function}
```
Comment (Simn) [inline HaxeManual/12-target-details.tex]: Did you check if that generates to markdown correctly?

**Comment by Simn** [comment]
@RealyUniqueName: Can you check the null-safety section?

File: HaxeManual/08-compiler-features.tex [hunk 17]
```diff
@@ -759,3 +759,148 @@ \section{Loop unrolling}
...
+	\item Out-of-bounds array writes fill all positions between the last defined index and the newly-written one with \expr{null} values. Null safety cannot protect against this.
+	\begin{lstlisting}
+var a:Array<String> = ["hello"];
+a[2] = "world";
+trace(a); // ["hello", null, "world"]
+var s:String = a[1]; // Cannot check this
+trace(s); //null
+\end{lstlisting}
+	\item Haxe was not designed with null safety in mind, so it's always possible \expr{null} values will come into your code from third-party code or even from the standard library.
+	\item Nullable fields and properties are not considered null-safe even after checking against \expr{null}. You can use helper methods instead (e.g. from \href{https://github.com/RealyUniqueName/Safety/}{this library}):
```
Comment (RealyUniqueName) [inline HaxeManual/08-compiler-features.tex]: I'm not sure it's ok to mention 3rd-party libraries in the manual.
Maybe it's better to provide an example code of such a helper method.

File: HaxeManual/12-target-details.tex (hunk 16 above)
Comment (Aurel300) [reply HaxeManual/12-target-details.tex]: Yup, it's fine.

File: HaxeManual/10-std.tex (hunk 15 above)
Comment (Simn) [reply HaxeManual/10-std.tex]: Read it again and nevermind, I'm a bit weird.

Comment (Gama11) [reply HaxeManual/10-std.tex]: Apparently I'm weird too then, I also found that a bit hard to parse.

Comment (Aurel300) [reply HaxeManual/10-std.tex]: Yeah that's a mistake, I accidentally a word.

> Due to threads having access to a shared memory space with all the Haxe variables and objects, it is possible ...

Is that better?

Comment (Simn) [reply HaxeManual/10-std.tex]: Much better.

File: HaxeManual/06-language-features.tex (hunk 7 above)
Comment (Aurel300) [reply HaxeManual/06-language-features.tex]: The short summary on the LF page or the entire section?

File: HaxeManual/10-std.tex (hunk 13 above)
Comment (Aurel300) [reply HaxeManual/10-std.tex]: https://github.com/HaxeFoundation/haxe/issues/8175

File: HaxeManual/06-language-features.tex (hunk 7 above)
Comment (Simn) [reply HaxeManual/06-language-features.tex]: I'd hold back on the entire section for now because Nicolas still wants to change something.
